Optimizations and WAs to support HPU execution for Detr-Resnet-50

sandeep-maddipatla opened the pull request:
Modifications to the Detr transformer including WA's and Optimizations to run the Detr-Resnet-50 model in eager and lazy modes on the HPU.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

sandeep-maddipatla commented:
This PR builds on #1155 , which is meant for Eager mode and adds changes necessary for Lazy mode execution. Some of the review feedback for the former PR is addressed here. More details below.

[x] pls. rebase/sync on top of main OH.
[x] run make style
[x] Pls. share the results of this test on g2 machines. 
    Done. Will share test result in another comment below.
[ ] we need to add README file for this in examples
    Skipped. There is a README.md at https://github.com/huggingface/optimum-habana/tree/main/examples/object-detection. We haven't changed that particular inference example. Pls let us know if that still needs modification.
[x] pls. add the appropriate CI tests for this.
     Done. Extended existing ci-test to add a `detr-resnet-50` test as well.

sandeep-maddipatla commented:
`make style` result:
`
![image](https://github.com/user-attachments/assets/078d6e71-ec70-494d-8dea-162478db032c)
`

sandeep-maddipatla commented:
Test Result:
![image](https://github.com/user-attachments/assets/61e1751c-c8b5-4f94-94f3-ca06022dee31)

vidyasiv commented on optimum/habana/transformers/models/detr/modeling_detr.py:
```diff
@@ -17,3 +29,181 @@ def gaudi_DetrConvModel_forward(self, pixel_values, pixel_mask):
+    """
+    Copied from https://github.com/huggingface/transformers/tree/v4.40.2
+    https://github.com/huggingface/transformers/blob/4fdf58afb72b0754da30037fc800b6044e7d9c99/src/transformers/models/detr/modeling_detr.py#L2287
+    The modications are:
```
```suggestion
    The modifications are:
```

vidyasiv commented on optimum/habana/transformers/models/detr/modeling_detr.py:
```diff
@@ -17,3 +29,181 @@ def gaudi_DetrConvModel_forward(self, pixel_values, pixel_mask):
+
+    # Compute the classification cost. Contrary to the loss, we don't use the NLL,
+    # but approximate it in 1 - proba[target class].
+    # The 1 is a constant that doesn't change the matching, it can be ommitted.
```
```suggestion
    # The 1 is a constant that doesn't change the matching, it can be omitted.
```

vidyasiv reviewed (changes requested):
Please rebase to latest main, there are some changes in modeling_utils.py

vidyasiv reviewed (changes requested):
@splotnikv , please take a look if you're covering for Sandeep

splotnikv commented:
> @splotnikv , please take a look if you're covering for Sandeep

Done. I don't have access right to update this PR, so created a new one. See https://github.com/huggingface/optimum-habana/pull/1404.

sandeep-maddipatla commented:
Rebased to latest optimum-habana, addressed feedback from #1404 , and merged in changes from that PR.

Now that I'm back working on this, will use this PR going forward to complete the review process. Sorry for the back-and-forth over the two PR's.

vidyasiv commented:
Sorry was not able to work last few days so unable to review sooner.
- Instead of jira link which shouldnt be pasted in public repo, can you add high level summary of the changes?
- Are the tests meant to address both lazy and eager modes? or should i be manually setting env to test that?
```
GAUDI2_CI=1
RUN_SLOW=true
#lazy mode all 8 pass but eager 4 fail
PT_HPU_LAZY_MODE=0 pytest tests/test_object_detection.py 
FAILED tests/test_object_detection.py::GaudiDETRTester::test_inference_hpu_graphs - AttributeError: module 'habana_frameworks.torch.hpu' has no attribute 'wrap_in_hpu_graph'
FAILED tests/test_object_detection.py::GaudiDETRTester::test_no_latency_regression_autocast - AttributeError: module 'habana_frameworks.torch.hpu' has no attribute 'wrap_in_hpu_graph'
FAILED tests/test_object_detection.py::GaudiDetrResnet50_Tester::test_inference_hpu_graphs - AttributeError: module 'habana_frameworks.torch.hpu' has no attribute 'wrap_in_hpu_graph'
FAILED tests/test_object_detection.py::GaudiDetrResnet50_Tester::test_no_latency_regression_autocast - AttributeError: module 'habana_frameworks.torch.hpu' has no attribute 'wrap_in_hpu_graph'
```
- README check (again not sure if eager mode is supported)
```
export PT_HPU_LAZY_MODE=0
python3 run_example.py \
	--model_name_or_path facebook/detr-resnet-101 \
	--image_path "http://images.cocodataset.org/val2017/000000039769.jpg" \
	--use_hpu_graphs \
	--bf16 \
	--print_result
```
```
AttributeError: module 'habana_frameworks.torch.hpu' has no attribute 'wrap_in_hpu_graph'
```
Lazy mode passes
```
Detected cat with confidence 0.996 at location [344.0, 25.25, 640.0, 376.0]
Detected remote with confidence 0.996 at location [328.0, 76.0, 372.0, 188.0]
Detected remote with confidence 0.996 at location [39.5, 69.5, 175.0, 119.0]
Detected cat with confidence 1.0 at location [15.62, 52.5, 316.0, 472.0]
Detected couch with confidence 0.996 at location [-1.25, 0.94, 640.0, 472.0]

Stats:
------------------------------------------------------------
Total latency (ms): 59.30161476135254 (for n_iterations=10) 
Average latency (ms): 5.930161476135254 (per iteration) 
```
Pls clarify how the testing is to be done.

vidyasiv commented:
@sandeep-maddipatla , could you update by EOW?

vidyasiv commented:
@sandeep-maddipatla  please resolve merge conflicts

emascarenhas commented:
@sandeep-maddipatla , Please make changes and post results of retesting, otherwise this will be pushed to the 1.20 release.

vidyasiv commented:
@sandeep-maddipatla  If we don't see an update by wed, this will need to be part of 1.20 release.

sandeep-maddipatla commented:
```
~/optimum-habana $ make style
...
ruff check . setup.py --fix
All checks passed!
ruff format . setup.py
401 files left unchanged
```


```
~/optimum-habana $ python setup.py install
~/optimum-habana $ pip install pytest timm sequencepiece
~/optimum-habana $ PT_HPU_LAZY_MODE=1 python -m pytest tests/test_object_detection.py
========================================================================================================================== test session starts ===========================================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ........                                                                                                                                                                                                                            [100%]

============================================================================================================================ warnings summary ============================================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================================ 8 passed, 2 warnings in 60.08s (0:01:00) ================================================================================================================


 ~/optimum-habana $ PT_HPU_LAZY_MODE=0 python -m pytest tests/test_object_detection.py
========================================================================================================================== test session starts ===========================================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ........                                                                                                                                                                                                                            [100%]

============================================================================================================================ warnings summary ============================================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpex/kernels/__init__.py:18
  /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpex/kernels/__init__.py:18: UserWarning: CustomNms, RoiAlignFunction, ScaledMaskedSoftmax from habana_frameworks.torch.hpex.kernels are no yet supported in eager mode
    warnings.warn(

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

tests/test_object_detection.py::GaudiDETRTester::test_inference_hpu_graphs
tests/test_object_detection.py::GaudiDetrResnet50_Tester::test_inference_hpu_graphs
  /root/optimum-habana/tests/test_object_detection.py:105: UserWarning: test_inference_hpu_graphs is supported only in lazy mode. Skipped
    warnings.warn("test_inference_hpu_graphs is supported only in lazy mode. Skipped")

tests/test_object_detection.py::GaudiDETRTester::test_no_latency_regression_autocast
tests/test_object_detection.py::GaudiDetrResnet50_Tester::test_no_latency_regression_autocast
  /root/optimum-habana/tests/test_object_detection.py:123: UserWarning: test_no_latency_regression_autocast is supported only in lazy mode. Skipped
    warnings.warn("test_no_latency_regression_autocast is supported only in lazy mode. Skipped")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================================================================== 8 passed, 7 warnings in 16.00s =====================================================================================================================
```

vidyasiv commented on tests/test_object_detection.py:
```diff
@@ -96,6 +101,9 @@ def test_inference_autocast(self):
             self.assertLess(np.abs(boxes[0].to(torch.float32).cpu().detach().numpy() - expected_location).max(), 5)
 
     def test_inference_hpu_graphs(self):
+        if is_eager_mode():
```
could you use pytest skipif decorator instead: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#id1

sandeep-maddipatla replied:
pushed another commit to implement this.

sandeep-maddipatla commented:
Test results with using the `skipIf` feature

```
~/optimum-habana $ PT_HPU_LAZY_MODE=1 python -m pytest tests/test_object_detection.py
========================================================================================================================== test session starts ===========================================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ........                                                                                                                                                                                                                            [100%]

============================================================================================================================ warnings summary ============================================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================================ 8 passed, 2 warnings in 63.79s (0:01:03) ================================================================================================================



~/optimum-habana $ PT_HPU_LAZY_MODE=0 python -m pytest tests/test_object_detection.py
========================================================================================================================== test session starts ===========================================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ..ss..ss                                                                                                                                                                                                                            [100%]

============================================================================================================================ warnings summary ============================================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpex/kernels/__init__.py:18
  /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpex/kernels/__init__.py:18: UserWarning: CustomNms, RoiAlignFunction, ScaledMaskedSoftmax from habana_frameworks.torch.hpex.kernels are no yet supported in eager mode
    warnings.warn(

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================================== 4 passed, 4 skipped, 3 warnings in 18.70s ================================================================================================================
```

vidyasiv commented:
@sandeep-maddipatla, could you resolve conflicts?

sandeep-maddipatla commented:
I did the rebase, but it appears the baseline commit (tip of `main` - commit `f48dda8`) now fails the style check. The same error that appears with main (shown below), also comes up with the rebased changes. These failures will not be addressed in this PR.

```
root@system-0001: ~/optimum-habana (main)$ make style
find . -name "habana_log.livealloc.log_*" -type f -delete
find . -name "hl-smi_log*" -type f -delete
find . -name .lock -type f -delete
find . -name .graph_dumps -type d -exec rm -r {} +
find . -name save-hpu.pdb -type f -delete
find . -name checkpoints.json -type f -delete
find . -name hpu_profile -type d -exec rm -r {} +
rm -rf regression/
rm -rf tmp_trainer/
rm -rf test/
rm -rf build/
rm -rf dist/
rm -rf optimum_habana.egg-info/
pip install -U pip ruff
Requirement already satisfied: pip in /usr/local/lib/python3.10/dist-packages (24.3.1)
Requirement already satisfied: ruff in /usr/local/lib/python3.10/dist-packages (0.9.1)
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
ruff check . setup.py --fix
All checks passed!
ruff format . setup.py
51 files reformatted, 360 files left unchanged

```

Test results look alright though.
```
$ PT_HPU_LAZY_MODE=1 python -m pytest tests/test_object_detection.py
========================================================================================================== test session starts ==========================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ........                                                                                                                                                                                           [100%]

=========================================================================================================== warnings summary ============================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================== 8 passed, 2 warnings in 65.60s (0:01:05) ================================================================================================

$ PT_HPU_LAZY_MODE=0 python -m pytest tests/test_object_detection.py
========================================================================================================== test session starts ==========================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /root/optimum-habana
configfile: setup.cfg
plugins: typeguard-4.3.0
collected 8 items

tests/test_object_detection.py ..ss..ss                                                                                                                                                                                           [100%]

=========================================================================================================== warnings summary ============================================================================================================
../../usr/lib/python3.10/inspect.py:288
  /usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    return isinstance(object, types.FunctionType)

../../usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24
  /usr/local/lib/python3.10/dist-packages/transformers-4.45.2-py3.10.egg/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================== 4 passed, 4 skipped, 2 warnings in 17.86s ===============================================================================================

```

vidyasiv reviewed (approved):
@regisss please review.
